DirectProgramming/DPC++FPGA/Tutorials/Features/loop_fusion/src/loop_fusion.cpp

```diff
@@ -0,0 +1,280 @@
+//==============================================================
+// Copyright Intel Corporation
+//
+// SPDX-License-Identifier: MIT
+// =============================================================
+#include <CL/sycl.hpp>
+#include <iomanip>
+#include <iostream>
+#include <sycl/ext/intel/fpga_extensions.hpp>
+
+// dpc_common.hpp can be found in the dev-utilities include folder.
+// e.g., $ONEAPI_ROOT/dev-utilities//include/dpc_common.hpp
+#include "dpc_common.hpp"
+
+constexpr size_t kN{1024};
+constexpr size_t kM{50};
+
+using namespace sycl;
+
+// Forward declare the kernel name in the global scope.
+// This FPGA best practice reduces name mangling in the optimization reports.
+class DefaultFusionKernel;
+class NoFusionKernel;
+class DefaultNoFusionKernel;
+class FusionFunctionKernel;
+
+#if defined(FPGA_EMULATOR)
+ext::intel::fpga_emulator_selector selector;
+#else
+ext::intel::fpga_selector selector;
+#endif
+
+// Handles error reporting
+void ErrorReport(sycl::exception const &e) {
+  // Catches exceptions in the host code
+  std::cerr << "Caught a SYCL host exception:\n" << e.what() << "\n";
+
+  // Most likely the runtime couldn't find FPGA hardware!
+  if (e.code().value() == CL_DEVICE_NOT_FOUND) {
+    std::cerr << "If you are targeting an FPGA, please ensure that your "
+                 "system has a correctly configured FPGA board.\n";
+    std::cerr << "Run sys_check in the oneAPI root directory to verify.\n";
+    std::cerr << "If you are targeting the FPGA emulator, compile with "
+                 "-DFPGA_EMULATOR.\n";
+  }
+  std::terminate();
+}
+
+// Returns kernel runtime in ns
+auto KernelRuntime(event e) {
+  auto start{e.get_profiling_info<info::event_profiling::command_start>()};
+  auto end{e.get_profiling_info<info::event_profiling::command_end>()};
+
+  return (end - start);
+}
+
+// Fuses inner loops by default, since the trip counts are equal, and there are
+// no dependencies between loops
+void DefaultFusion(std::array<int, kM> &m_array_1,
+                   std::array<int, kM> &m_array_2) {
+  try {
+    queue q(selector, dpc_common::exception_handler,
+            property::queue::enable_profiling{});
+
+    buffer buffer_array_1(m_array_1);
```

Comment:
The naming of the array, buffer, and accessor could be cleaned up here.    Normally we want the compute loop to be the simplest/cleanest representation.   This then causes the buffer to be more verbose/different and the original data array to be also different.   Think of it as trying to minimize the most frequently used identifiers and emphasize the most frequently executed as also simplest (i.e. inner loops).

Reply:
Cleaner meaning shorter in this case? I see in other code samples, buffers named a_buf and the like; could I name my buffers buf_1, buf_2?

Reply:
I think Paul would like names like this:
array_a (origianl array)
buf_a (for the buffer)
a (inside the kernel)

So the name used in the kernel is as simple as possible.